pilot/pkg/model/push_context.go

```diff
@@ -284,15 +286,27 @@ func (first *PushRequest) Merge(other *PushRequest) *PushRequest {
 	// Only merge EdsUpdates when incremental eds push needed.
 	if !merged.Full {
 		merged.EdsUpdates = make(map[string]struct{})
+
 		// Merge the updates
 		for update := range first.EdsUpdates {
 			merged.EdsUpdates[update] = struct{}{}
 		}
 		for update := range other.EdsUpdates {
 			merged.EdsUpdates[update] = struct{}{}
 		}
+
+		if first.EgdsUpdates != nil || other.EgdsUpdates != nil {
+			merged.EgdsUpdates = make(map[string]struct{})
+			for update := range first.EgdsUpdates {
+				merged.EgdsUpdates[update] = struct{}{}
+			}
+			for update := range other.EgdsUpdates {
+				merged.EgdsUpdates[update] = struct{}{}
+			}
+		}
```

Comment:
Why do you need to have another copy same state in the form of EdgsUpdates? Why can't you use the existing EdsUpdates ?

Reply:
The 'EgdsUpdates' stores the changed groups while the 'EdsUpdates' only has the changed service. For example, if we a service named A has 100 endpoints and we slice it into 5 groups. With 15 endpoints change will probably involve 2 or more groups.  So the data in 'EdsUpdates' will be '{A: {}}' while 'EgdsUpdates' will be {'A-G1':{},'A-G2':{},'A-G5':{}}.

> Why do you need to have another copy same state in the form of EdgsUpdates? Why can't you use the existing EdsUpdates ?

The 'EgdsUpdates' stores the changed groups while the 'EdsUpdates' only has the changed service. For example, if we a service named A has 100 endpoints and we slice it into 5 groups. With 15 endpoints change will probably involve 2 or more groups.  So the data in 'EdsUpdates' will be '{A: {}}' while 'EgdsUpdates' will be {'A-G1':{},'A-G2':{},'A-G5':{}}.